diff --git a/src/ipc-dbus/DBusIPCAPI_v1.cpp b/src/ipc-dbus/DBusIPCAPI_v1.cpp
index 1ac9160..e39f70b 100644
--- a/src/ipc-dbus/DBusIPCAPI_v1.cpp
+++ b/src/ipc-dbus/DBusIPCAPI_v1.cpp
@@ -43,6 +43,7 @@
 
 #include "DBUSHelpers.h"
 #include "any-to.h"
+#include "commissioner-utils.h"
 
 using namespace DBUSHelpers;
 using namespace nl;
@@ -78,6 +79,8 @@ DBusIPCAPI_v1::init_callback_tables()
 	INTERFACE_CALLBACK_CONNECT(WPANTUND_IF_CMD_ROUTE_ADD, interface_route_add_handler);
 	INTERFACE_CALLBACK_CONNECT(WPANTUND_IF_CMD_ROUTE_REMOVE, interface_route_remove_handler);
 
+	INTERFACE_CALLBACK_CONNECT(WPANTUND_IF_CMD_JOINER_ADD, interface_joiner_add_handler);
+
 	INTERFACE_CALLBACK_CONNECT(WPANTUND_IF_CMD_DATA_POLL, interface_data_poll_handler);
 	INTERFACE_CALLBACK_CONNECT(WPANTUND_IF_CMD_CONFIG_GATEWAY, interface_config_gateway_handler);
 
@@ -1192,6 +1195,57 @@ bail:
 	return ret;
 }
 
+DBusHandlerResult
+DBusIPCAPI_v1::interface_joiner_add_handler(
+   NCPControlInterface* interface,
+   DBusMessage *        message
+) {
+	DBusHandlerResult ret = DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+
+	dbus_message_ref(message);
+
+	const uint8_t* ext_addr = NULL;
+	int ext_addr_len = 0;
+	const char* psk = NULL;
+	int psk_len = 0;
+	uint32_t joiner_timeout = 0;
+	bool did_succeed = false;
+
+	did_succeed = dbus_message_get_args(
+		message, NULL,
+		DBUS_TYPE_STRING, &psk,
+		DBUS_TYPE_UINT32, &joiner_timeout,
+		DBUS_TYPE_ARRAY, DBUS_TYPE_BYTE, &ext_addr, &ext_addr_len,
+		DBUS_TYPE_INVALID
+	);
+
+	if (!did_succeed) {
+		// No extended address specified
+		did_succeed = dbus_message_get_args(
+			message, NULL,
+			DBUS_TYPE_STRING, &psk,
+			DBUS_TYPE_UINT32, &joiner_timeout,
+			DBUS_TYPE_INVALID
+		);
+    }
+
+	require(did_succeed, bail);
+	require(psk != NULL, bail);
+
+	dbus_message_ref(message);
+	interface->joiner_add(
+		psk,
+		joiner_timeout,
+		ext_addr,
+		boost::bind(&DBusIPCAPI_v1::CallbackWithStatus_Helper, this, _1, message)
+	);
+
+	ret = DBUS_HANDLER_RESULT_HANDLED;
+
+bail:
+
+	return ret;
+}
 
 DBusHandlerResult
 DBusIPCAPI_v1::message_handler(
diff --git a/src/ipc-dbus/DBusIPCAPI_v1.h b/src/ipc-dbus/DBusIPCAPI_v1.h
index 79cf3a1..6827955 100644
--- a/src/ipc-dbus/DBusIPCAPI_v1.h
+++ b/src/ipc-dbus/DBusIPCAPI_v1.h
@@ -188,6 +188,11 @@ private:
 		DBusMessage *        message
 	);
 
+	DBusHandlerResult interface_joiner_add_handler(
+		NCPControlInterface* interface,
+		DBusMessage *        message
+	);
+
 private:
 	typedef DBusHandlerResult (interface_handler_cb)(
 		NCPControlInterface*,
diff --git a/src/ipc-dbus/wpan-dbus-v1.h b/src/ipc-dbus/wpan-dbus-v1.h
index 2d573d8..f8eff4b 100644
--- a/src/ipc-dbus/wpan-dbus-v1.h
+++ b/src/ipc-dbus/wpan-dbus-v1.h
@@ -76,6 +76,8 @@
 #define WPANTUND_IF_CMD_PROP_SET              "PropSet"
 #define WPANTUND_IF_SIGNAL_PROP_CHANGED       "PropChanged"
 
+#define WPANTUND_IF_CMD_JOINER_ADD             "JoinerAdd"
+
 // ============================================================================
 // NestLabs Internal API Interface
 
diff --git a/src/ncp-dummy/DummyNCPControlInterface.cpp b/src/ncp-dummy/DummyNCPControlInterface.cpp
index 518a70a..cf0b0cd 100644
--- a/src/ncp-dummy/DummyNCPControlInterface.cpp
+++ b/src/ncp-dummy/DummyNCPControlInterface.cpp
@@ -169,13 +169,23 @@ DummyNCPControlInterface::remove_external_route(
 }
 
 void
+DummyNCPControlInterface::joiner_add(
+	const char *psk,
+	uint32_t joiner_timeout,
+	const uint8_t *addr,
+	CallbackWithStatus cb
+) {
+	cb(kWPANTUNDStatus_FeatureNotImplemented);
+}
+
+void
 DummyNCPControlInterface::permit_join(
     int seconds,
     uint8_t traffic_type,
     in_port_t traffic_port,
     bool network_wide,
     CallbackWithStatus cb
-    )
+)
 {
 	// TODO: Writeme!
 	cb(kWPANTUNDStatus_FeatureNotImplemented);
diff --git a/src/ncp-dummy/DummyNCPControlInterface.h b/src/ncp-dummy/DummyNCPControlInterface.h
index 5bac656..a898cd9 100644
--- a/src/ncp-dummy/DummyNCPControlInterface.h
+++ b/src/ncp-dummy/DummyNCPControlInterface.h
@@ -110,6 +110,13 @@ public:
 		CallbackWithStatus cb = NilReturn()
 	);
 
+	virtual void joiner_add(
+		const char *psk,
+		uint32_t joiner_timeout,
+		const uint8_t *addr,
+		CallbackWithStatus cb = NilReturn()
+	);
+
 	virtual void data_poll(CallbackWithStatus cb = NilReturn());
 	virtual void host_did_wake(CallbackWithStatus cb = NilReturn());
 
diff --git a/src/ncp-spinel/SpinelNCPControlInterface.cpp b/src/ncp-spinel/SpinelNCPControlInterface.cpp
index c8a5df3..f711507 100644
--- a/src/ncp-spinel/SpinelNCPControlInterface.cpp
+++ b/src/ncp-spinel/SpinelNCPControlInterface.cpp
@@ -32,6 +32,7 @@
 #include "string-utils.h"
 #include "any-to.h"
 #include "time-utils.h"
+#include "commissioner-utils.h"
 
 #include <cstring>
 #include <algorithm>
@@ -309,6 +310,56 @@ bail:
 }
 
 void
+SpinelNCPControlInterface::joiner_add(
+		const char *psk,
+		uint32_t joiner_timeout,
+		const uint8_t *addr,
+		CallbackWithStatus cb
+) {
+
+	require_action(psk != NULL, bail, cb(kWPANTUNDStatus_InvalidArgument));
+	require_action(mNCPInstance->mEnabled, bail, cb(kWPANTUNDStatus_InvalidWhenDisabled));
+
+	if (addr) {
+		mNCPInstance->start_new_task(SpinelNCPTaskSendCommand::Factory(mNCPInstance)
+			.set_callback(cb)
+			.add_command(SpinelPackData(
+				SPINEL_FRAME_PACK_CMD_PROP_VALUE_INSERT(
+					SPINEL_DATATYPE_UTF8_S
+					SPINEL_DATATYPE_UINT32_S
+					SPINEL_DATATYPE_EUI64_S
+				),
+				SPINEL_PROP_THREAD_JOINERS,
+				psk,
+				joiner_timeout,
+				addr
+			))
+			.set_lock_property(SPINEL_PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE)
+			.finish()
+		);
+	}
+	else {
+		mNCPInstance->start_new_task(SpinelNCPTaskSendCommand::Factory(mNCPInstance)
+			.set_callback(cb)
+			.add_command(SpinelPackData(
+				SPINEL_FRAME_PACK_CMD_PROP_VALUE_INSERT(
+					SPINEL_DATATYPE_UTF8_S
+					SPINEL_DATATYPE_UINT32_S
+				),
+				SPINEL_PROP_THREAD_JOINERS,
+				psk,
+				joiner_timeout
+			))
+			.set_lock_property(SPINEL_PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE)
+			.finish()
+		);
+	}
+
+bail:
+	return;
+}
+
+void
 SpinelNCPControlInterface::remove_external_route(
 	const struct in6_addr *prefix,
 	int prefix_len_in_bits,
diff --git a/src/ncp-spinel/SpinelNCPControlInterface.h b/src/ncp-spinel/SpinelNCPControlInterface.h
index b91335a..0411b8a 100644
--- a/src/ncp-spinel/SpinelNCPControlInterface.h
+++ b/src/ncp-spinel/SpinelNCPControlInterface.h
@@ -109,6 +109,13 @@ public:
 		CallbackWithStatus cb = NilReturn()
 	);
 
+	virtual void joiner_add(
+		const char *psk,
+		uint32_t joiner_timeout,
+		const uint8_t *addr,
+		CallbackWithStatus cb = NilReturn()
+	);
+
 	virtual void data_poll(CallbackWithStatus cb = NilReturn());
 	virtual void host_did_wake(CallbackWithStatus cb = NilReturn());
 
diff --git a/src/ncp-spinel/SpinelNCPInstance.cpp b/src/ncp-spinel/SpinelNCPInstance.cpp
index 8f6a3a3..88e107f 100644
--- a/src/ncp-spinel/SpinelNCPInstance.cpp
+++ b/src/ncp-spinel/SpinelNCPInstance.cpp
@@ -247,6 +247,7 @@ SpinelNCPInstance::get_supported_property_keys()const
 		properties.insert(kWPANTUNDProperty_ThreadStableLeaderNetworkData);
 		properties.insert(kWPANTUNDProperty_ThreadChildTable);
 		properties.insert(kWPANTUNDProperty_ThreadNeighborTable);
+		properties.insert(kWPANTUNDProperty_ThreadCommissionerEnabled);
 	}
 
 	if (mCapabilities.count(SPINEL_CAP_COUNTERS)) {
@@ -351,6 +352,7 @@ static int unpack_jam_detect_history_bitmap(const uint8_t *data_in, spinel_size_
 {
 	spinel_ssize_t len;
 	uint32_t lower, higher;
+	uint64_t val;
 	int ret = kWPANTUNDStatus_Failure;
 
 	len = spinel_datatype_unpack(
@@ -450,6 +452,9 @@ SpinelNCPInstance::get_property(
 	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_ThreadStableNetworkDataVersion)) {
 		SIMPLE_SPINEL_GET(SPINEL_PROP_THREAD_STABLE_NETWORK_DATA_VERSION, SPINEL_DATATYPE_UINT8_S);
 
+	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_ThreadCommissionerEnabled)) {
+		SIMPLE_SPINEL_GET(SPINEL_PROP_THREAD_COMMISSIONER_ENABLED, SPINEL_DATATYPE_BOOL_S);
+
 	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_IPv6MeshLocalPrefix) && !buffer_is_nonzero(mNCPV6Prefix, sizeof(mNCPV6Prefix))) {
 		SIMPLE_SPINEL_GET(SPINEL_PROP_IPV6_ML_PREFIX, SPINEL_DATATYPE_IPv6ADDR_S);
 
@@ -947,6 +952,18 @@ SpinelNCPInstance::set_property(
 				);
 			}
 
+		} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_ThreadCommissionerEnabled)) {
+			bool isEnabled = any_to_bool(value);
+
+			start_new_task(SpinelNCPTaskSendCommand::Factory(this)
+				.set_callback(cb)
+				.add_command(SpinelPackData(
+					SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_BOOL_S),
+					SPINEL_PROP_THREAD_COMMISSIONER_ENABLED,
+					isEnabled
+				))
+				.finish()
+			);
 		} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_OpenThreadLogLevel)) {
 			uint8_t logLevel = static_cast<uint8_t>(any_to_int(value));
 			Data command = SpinelPackData(SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_UINT8_S), SPINEL_PROP_DEBUG_NCP_LOG_LEVEL, logLevel);
diff --git a/src/util/Makefile.am b/src/util/Makefile.am
index 15dc99d..ef0070c 100644
--- a/src/util/Makefile.am
+++ b/src/util/Makefile.am
@@ -66,6 +66,7 @@ EXTRA_DIST = \
 	Timer.cpp \
 	sec-random.h \
 	sec-random.c \
+	commissioner-utils.h \
 	$(NULL)
 
 DISTCLEANFILES = \
diff --git a/src/util/commissioner-utils.h b/src/util/commissioner-utils.h
new file mode 100644
index 0000000..3e47f89
--- /dev/null
+++ b/src/util/commissioner-utils.h
@@ -0,0 +1,30 @@
+/*
+ *
+ * Copyright (c) 2017 OpenThread Authors, Inc.
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef COMMISSIONER_UTILS_H
+#define COMMISSIONER_UTILS_H
+
+#define PSK_MIN_LENGTH              6
+#define PSK_MAX_LENGTH              32
+#define INVALID_PSK_CHARACTERS      "IOQZ"
+
+#define EXT_ADDRESS_LENGTH          8
+#define EXT_ADDRESS_LENGTH_CHAR     (2*EXT_ADDRESS_LENGTH)
+#define DEFAULT_JOINER_TIMEOUT      120
+#endif
diff --git a/src/util/string-utils.c b/src/util/string-utils.c
index d3f13c3..3bb8145 100644
--- a/src/util/string-utils.c
+++ b/src/util/string-utils.c
@@ -297,3 +297,26 @@ buffer_is_nonzero(const uint8_t* buffer, size_t len)
 	}
 	return false;
 }
+
+bool
+is_hex(const uint8_t* buff, size_t len)
+{
+	while (len--) {
+		if (!isxdigit(*buff++)) {
+			return false;
+		}
+	}
+	return true;
+}
+
+bool
+is_uppercase_or_digit(const uint8_t* buff, size_t len)
+{
+	while (len--) {
+		if (!(isupper(*buff) || isdigit(*buff))) {
+			return false;
+		}
+		buff++;
+	}
+	return true;
+}
diff --git a/src/util/string-utils.h b/src/util/string-utils.h
index 528fbd2..81fb34b 100644
--- a/src/util/string-utils.h
+++ b/src/util/string-utils.h
@@ -42,6 +42,8 @@ extern int parse_string_into_data(uint8_t* buffer, size_t len, const char* c_str
 extern int encode_data_into_string(const uint8_t*  buffer, size_t len, char* c_str, size_t c_str_max_len, int pad_to);
 extern int strtologmask(const char* value, int prev_mask);
 extern bool buffer_is_nonzero(const uint8_t* buffer, size_t len);
+extern bool is_hex(const uint8_t* buff, size_t len);
+extern bool is_uppercase_or_digit(const uint8_t* buff, size_t len);
 
 static inline char
 int_to_hex_digit(uint8_t x)
diff --git a/src/wpanctl/Makefile.am b/src/wpanctl/Makefile.am
index 552ef3b..87dd611 100644
--- a/src/wpanctl/Makefile.am
+++ b/src/wpanctl/Makefile.am
@@ -78,6 +78,8 @@ wpanctl_SOURCES = \
 	tool-cmd-remove-route.h \
 	tool-cmd-pcap.h \
 	tool-cmd-pcap.c \
+	tool-cmd-commissioner.h \
+	tool-cmd-commissioner.c \
 	$(NULL)
 
 wpanctl_LDADD = \
diff --git a/src/wpanctl/tool-cmd-commissioner.c b/src/wpanctl/tool-cmd-commissioner.c
new file mode 100644
index 0000000..853bd0a
--- /dev/null
+++ b/src/wpanctl/tool-cmd-commissioner.c
@@ -0,0 +1,369 @@
+/*
+ *
+ * Copyright (c) 2017 OpenThread Authors, Inc.
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *    Description:
+ *      This file implements "commissioner" command in wpanctl.
+ *
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <getopt.h>
+#include "wpanctl-utils.h"
+#include "tool-cmd-setprop.h"
+#include "assert-macros.h"
+#include "args.h"
+#include "assert-macros.h"
+#include "wpan-dbus-v1.h"
+#include "string-utils.h"
+#include "commissioner-utils.h"
+
+const char commissioner_cmd_syntax[] = "[args] <psk> [address] [joiner_timeout [s]]";
+
+static const arg_list_item_t commissioner_option_list[] = {
+    {'h', "help", NULL, "Print Help"},
+    {'t', "timeout", "ms", "Set timeout period"},
+    {'e', "start", NULL, "Start native commissioner"},
+    {'d', "stop", NULL, "Stop native commissioner"},
+    {'a', "joiner-add", NULL, "Add joiner"},
+    {'r', "joiner-remove", NULL, "Remove joiner"},
+    {'s', "state", NULL, "State information"},
+    {0}
+};
+
+int tool_cmd_commissioner(int argc, char* argv[])
+{
+    int ret = 0;
+    int c;
+    int timeout = DEFAULT_TIMEOUT_IN_SECONDS * 1000;
+    DBusConnection* connection = NULL;
+    DBusMessage *message = NULL;
+    DBusMessage *reply = NULL;
+    DBusError error;
+    const char* property_name = NULL;
+    const char* ext_addr = NULL;
+    const char* psk = NULL;
+    int psk_len = 0;
+    uint32_t joiner_timeout = DEFAULT_JOINER_TIMEOUT;
+    dbus_bool_t enabled = false;
+    const char* invalid_psk_characters = INVALID_PSK_CHARACTERS;
+    const char* property_commissioner_enabled = kWPANTUNDProperty_ThreadCommissionerEnabled;
+    const char* property_commissioner_enabled_value = "false";
+
+    char path[DBUS_MAXIMUM_NAME_LENGTH+1];
+    char interface_dbus_name[DBUS_MAXIMUM_NAME_LENGTH+1];
+
+    dbus_error_init(&error);
+    while (1) {
+        static struct option long_options[] = {
+            {"help", no_argument, 0, 'h'},
+            {"timeout", required_argument, 0, 't'},
+            {"start", no_argument, 0, 'e'},
+            {"stop", no_argument, 0, 'd'},
+            {"joiner-add", no_argument, 0, 'a'},
+            {"remove", required_argument, 0, 'r'},
+            {"state", no_argument, 0, 's'},
+            {0, 0, 0, 0}
+        };
+
+        int option_index = 0;
+        c = getopt_long(argc, argv, "hst:edr:a", long_options,
+                        &option_index);
+        if (c == -1)
+            break;
+
+        switch (c) {
+        case 'h':
+            print_arg_list_help(commissioner_option_list,
+                                argv[0], commissioner_cmd_syntax);
+            ret = ERRORCODE_HELP;
+            goto bail;
+
+        case 's':
+            //state
+            ret = ERRORCODE_NOT_IMPLEMENTED;
+            goto bail;
+
+        case 't':
+            //timeout
+            timeout = strtol(optarg, NULL, 0);
+            break;
+
+        case 'e':
+            // start (enabled)
+            property_commissioner_enabled_value = "true";
+            // intentionally pass through
+        case 'd':
+            //stop (disabled)
+            ret = lookup_dbus_name_from_interface(interface_dbus_name, gInterfaceName);
+            if (gInterfaceName[0] == 0) {
+                fprintf(stderr,
+                        "%s: error: No WPAN interface set (use the `cd` command, or the `-I` argument for `wpanctl`).\n",
+                        argv[0]);
+                ret = ERRORCODE_BADARG;
+                goto bail;
+            }
+            connection = dbus_bus_get(DBUS_BUS_STARTER, &error);
+
+            if (!connection) {
+                dbus_error_free(&error);
+                dbus_error_init(&error);
+                connection = dbus_bus_get(DBUS_BUS_SYSTEM, &error);
+            }
+            require_string(connection != NULL, bail, error.message);
+
+            if (ret != 0) {
+                print_error_diagnosis(ret);
+                goto bail;
+            }
+            snprintf(path,
+                     sizeof(path),
+                     "%s/%s",
+                     WPANTUND_DBUS_PATH,
+                     gInterfaceName);
+
+            message = dbus_message_new_method_call(
+                interface_dbus_name,
+                path,
+                WPANTUND_DBUS_APIv1_INTERFACE,
+                WPANTUND_IF_CMD_PROP_SET
+                );
+
+            dbus_message_append_args(
+                message,
+                DBUS_TYPE_STRING, &property_commissioner_enabled,
+                DBUS_TYPE_INVALID
+                );
+
+            dbus_message_append_args(
+                message,
+                DBUS_TYPE_STRING, &property_commissioner_enabled_value,
+                DBUS_TYPE_INVALID
+                );
+
+            reply = dbus_connection_send_with_reply_and_block(
+                connection,
+                message,
+                timeout,
+                &error
+                );
+
+            if (!reply) {
+                fprintf(stderr, "%s: error: %s\n", argv[0], error.message);
+                ret = ERRORCODE_TIMEOUT;
+                goto bail;
+            }
+
+            dbus_message_get_args(reply, &error,
+                                  DBUS_TYPE_INT32, &ret,
+                                  DBUS_TYPE_INVALID
+                                  );
+
+            if (!ret) {
+                fprintf(stderr, "Commissioner command applied.\n");
+            } else {
+                fprintf(stderr, "%s failed with error %d. %s\n", argv[0], ret, wpantund_status_to_cstr(ret));
+                print_error_diagnosis(ret);
+            }
+
+            goto bail;
+
+        case 'r':
+            // remove
+            ret = ERRORCODE_NOT_IMPLEMENTED;
+            goto bail;
+
+        case 'a':
+            // add
+            if (optind < argc) {
+                if (!psk) {
+                    psk = argv[optind];
+                    psk_len = strnlen(psk, PSK_MAX_LENGTH+1);
+                    optind++;
+                }
+            }
+
+            if (optind < argc) {
+                if (!ext_addr) {
+                    ext_addr = argv[optind];
+                    optind++;
+                }
+            }
+
+            if (optind < argc) {
+                joiner_timeout = (uint32_t) strtol(argv[optind], NULL, 0);
+                optind++;
+            }
+
+            if (optind < argc) {
+                fprintf(stderr,
+                        "%s: error: Unexpected extra argument: \"%s\"\n",
+                        argv[0], argv[optind]);
+                ret = ERRORCODE_BADARG;
+                goto bail;
+            }
+
+            if (!ext_addr)
+                fprintf(stderr, "%s: warning: No address value specified, any joiner knowing PSKd can join.\n", argv[0]);
+            else {
+                if (!is_hex(ext_addr, EXT_ADDRESS_LENGTH_CHAR)) {
+                    fprintf(stderr, "%s: error: Invalid address.\n", argv[0]);
+                    ret = ERRORCODE_BADARG;
+                    goto bail;
+                }
+                if (strnlen(ext_addr, EXT_ADDRESS_LENGTH_CHAR+1) != EXT_ADDRESS_LENGTH_CHAR) {
+                    fprintf(stderr, "%s: error: Wrong address length.%d \n", argv[0], strnlen(ext_addr, EXT_ADDRESS_LENGTH_CHAR+1));
+                    ret = ERRORCODE_BADARG;
+                    goto bail;
+                }
+            }
+
+            if (!psk) {
+                fprintf(stderr, "%s: error: Missing PSK value.\n", argv[0]);
+                ret = ERRORCODE_BADARG;
+                goto bail;
+            }
+
+            if (psk_len < PSK_MIN_LENGTH || psk_len > PSK_MAX_LENGTH) {
+                fprintf(stderr, "%s: error: Invalid PSK length.\n", argv[0]);
+                ret = ERRORCODE_BADARG;
+                goto bail;
+            }
+
+            if (strpbrk(invalid_psk_characters, psk) != 0) {
+                fprintf(stderr, "%s: warning: PSK consists an invalid character\n", argv[0]);
+            }
+
+            if (!is_uppercase_or_digit(psk, psk_len)) {
+                fprintf(stderr, "%s: warning: PSK should consist only uppercase letters and digits\n", argv[0]);
+            }
+
+            if (gInterfaceName[0] == 0) {
+                fprintf(stderr,
+                        "%s: error: No WPAN interface set (use the `cd` command, or the `-I` argument for `wpanctl`).\n",
+                        argv[0]);
+                ret = ERRORCODE_BADARG;
+                goto bail;
+            }
+            connection = dbus_bus_get(DBUS_BUS_STARTER, &error);
+
+            if (!connection) {
+                dbus_error_free(&error);
+                dbus_error_init(&error);
+                connection = dbus_bus_get(DBUS_BUS_SYSTEM, &error);
+            }
+
+            require_string(connection != NULL, bail, error.message);
+
+            ret = lookup_dbus_name_from_interface(interface_dbus_name, gInterfaceName);
+            if (ret != 0) {
+                print_error_diagnosis(ret);
+                goto bail;
+            }
+            snprintf(path,
+                     sizeof(path),
+                     "%s/%s",
+                     WPANTUND_DBUS_PATH,
+                     gInterfaceName);
+
+            message = dbus_message_new_method_call(
+                interface_dbus_name,
+                path,
+                WPANTUND_DBUS_APIv1_INTERFACE,
+                WPANTUND_IF_CMD_JOINER_ADD
+                );
+
+            if (psk) {
+                uint8_t psk_bytes[psk_len];
+                memset(psk_bytes, '\0', psk_len+1);
+
+                memcpy(psk_bytes, psk, psk_len);
+                char *psk = psk_bytes;
+
+                dbus_message_append_args(
+                    message,
+                    DBUS_TYPE_STRING, &psk,
+                    DBUS_TYPE_INVALID
+                    );
+
+                dbus_message_append_args(
+                    message,
+                    DBUS_TYPE_UINT32, &joiner_timeout,
+                    DBUS_TYPE_INVALID
+                    );
+
+                if (ext_addr) {
+                    uint8_t addr_bytes[EXT_ADDRESS_LENGTH];
+                    memset(addr_bytes, 0, EXT_ADDRESS_LENGTH);
+                    uint8_t *p_addr_bytes = addr_bytes;
+                    int length = parse_string_into_data(addr_bytes,
+                                                        EXT_ADDRESS_LENGTH,
+                                                        ext_addr);
+                    assert(length == EXT_ADDRESS_LENGTH);
+                    dbus_message_append_args(
+                        message,
+                        DBUS_TYPE_ARRAY, DBUS_TYPE_BYTE, &p_addr_bytes, length,
+                        DBUS_TYPE_INVALID
+                        );
+                }
+
+                reply = dbus_connection_send_with_reply_and_block(
+                    connection,
+                    message,
+                    timeout,
+                    &error
+                    );
+
+                fprintf(stderr, "After reply joiner\n");
+                if (!reply) {
+                    fprintf(stderr, "%s: error: %s\n", argv[0], error.message);
+                    ret = ERRORCODE_TIMEOUT;
+                    goto bail;
+                }
+
+                fprintf(stderr, "Get reply joiner timeout\n");
+                dbus_message_get_args(reply, &error,
+                                      DBUS_TYPE_INT32, &ret,
+                                      DBUS_TYPE_INVALID
+                                      );
+                if (!ret) {
+                    fprintf(stderr, "Joiner added.\n");
+                } else {
+                    fprintf(stderr, "%s failed with error %d. %s\n", argv[0], ret, wpantund_status_to_cstr(ret));
+                    print_error_diagnosis(ret);
+                }
+            }
+        }
+    }
+
+
+bail:
+    if (connection)
+        dbus_connection_unref(connection);
+
+    if (message)
+        dbus_message_unref(message);
+
+    if (reply)
+        dbus_message_unref(reply);
+
+    dbus_error_free(&error);
+
+    return ret;
+}
diff --git a/src/wpanctl/tool-cmd-commissioner.h b/src/wpanctl/tool-cmd-commissioner.h
new file mode 100644
index 0000000..6fa128b
--- /dev/null
+++ b/src/wpanctl/tool-cmd-commissioner.h
@@ -0,0 +1,30 @@
+/*
+ *
+ * Copyright (c) 2017 OpenThread Authors, Inc.
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *    Description:
+ *      This file implements "commissioner" command in wpanctl.
+ *
+ */
+
+#ifndef WPANCTL_TOOL_CMD_COMMISSIONER_H
+#define WPANCTL_TOOL_CMD_COMMISSIONER_H
+
+#include "wpanctl-utils.h"
+
+int tool_cmd_commissioner(int argc, char* argv[]);
+
+#endif // WPANCTL_TOOL_CMD_COMMISSIONER_H
diff --git a/src/wpanctl/wpanctl-cmds.h b/src/wpanctl/wpanctl-cmds.h
index d738cda..c80549e 100644
--- a/src/wpanctl/wpanctl-cmds.h
+++ b/src/wpanctl/wpanctl-cmds.h
@@ -41,6 +41,7 @@
 #include "tool-cmd-add-route.h"
 #include "tool-cmd-remove-route.h"
 #include "tool-cmd-pcap.h"
+#include "tool-cmd-commissioner.h"
 
 #include "wpanctl-utils.h"
 
@@ -101,6 +102,11 @@
 		&tool_cmd_remove_route \
 	}, \
 	{ \
+		"commissioner", \
+		"Commissioner commands", \
+		&tool_cmd_commissioner \
+	}, \
+	{ \
 		"list", \
 		"List available interfaces.", \
 		&tool_cmd_list \
diff --git a/src/wpantund/NCPControlInterface.h b/src/wpantund/NCPControlInterface.h
index cef4e84..576fa0f 100644
--- a/src/wpantund/NCPControlInterface.h
+++ b/src/wpantund/NCPControlInterface.h
@@ -103,7 +103,6 @@ public:
 
 	virtual void leave(CallbackWithStatus cb = NilReturn()) = 0;
 
-	//! Deprecated. Set kWPANTUNDProperty_InterfaceUp to true instead.
 	virtual void attach(CallbackWithStatus cb = NilReturn()) = 0;
 
 	virtual void reset(CallbackWithStatus cb = NilReturn()) = 0;
@@ -140,6 +139,13 @@ public:
 		CallbackWithStatus cb = NilReturn()
 	) = 0;
 
+	virtual void joiner_add(
+		const char *psk,
+		uint32_t joiner_timeout,
+		const uint8_t *addr,
+		CallbackWithStatus cb = NilReturn()
+	) = 0;
+
 	virtual void remove_external_route(
 		const struct in6_addr *prefix,
 		int prefix_len_in_bits,
@@ -205,6 +211,7 @@ public:
 	    CallbackWithStatus cb = NilReturn()
 	) = 0;
 
+
 public:
 	// ========================================================================
 	// Convenience methods
@@ -223,6 +230,7 @@ public:
 
 	static bool translate_deprecated_property(std::string& key);
 
+
 public:
 	// ========================================================================
 	// Signals
diff --git a/src/wpantund/wpan-properties.h b/src/wpantund/wpan-properties.h
index 8dd8b9d..d0807bf 100644
--- a/src/wpantund/wpan-properties.h
+++ b/src/wpantund/wpan-properties.h
@@ -96,6 +96,7 @@
 #define kWPANTUNDProperty_ThreadStableNetworkData        "Thread:StableNetworkData"
 #define kWPANTUNDProperty_ThreadStableNetworkDataVersion "Thread:StableNetworkDataVersion"
 #define kWPANTUNDProperty_ThreadPreferredRouterID        "Thread:PreferredRouterID"
+#define kWPANTUNDProperty_ThreadCommissionerEnabled      "Thread:Commissioner:Enabled"
 
 #define kWPANTUNDProperty_OpenThreadLogLevel                   "OpenThread:LogLevel"
 #define kWPANTUNDProperty_OpenThreadMsgBufferCounters          "OpenThread:MsgBufferCounters"
diff --git a/third_party/openthread/src/ncp/spinel.h b/third_party/openthread/src/ncp/spinel.h
index 36e64a4..9c1a200 100644
--- a/third_party/openthread/src/ncp/spinel.h
+++ b/third_party/openthread/src/ncp/spinel.h
@@ -797,7 +797,7 @@ typedef enum
     /** Format: `C`
      */
     SPINEL_PROP_THREAD_CHILD_COUNT_MAX = SPINEL_PROP_THREAD_EXT__BEGIN + 12,
-
+   
     /// Leader network data
     /** Format: `D` - Read only
      */
@@ -810,6 +810,18 @@ typedef enum
     SPINEL_PROP_THREAD_STABLE_LEADER_NETWORK_DATA
                                        = SPINEL_PROP_THREAD_EXT__BEGIN + 14,
 
+    /// Thread joiner data
+    /** Format `A(T(ULE))`
+    *  PSKd, joiner timeout, eui64 (optional)
+    */
+    SPINEL_PROP_THREAD_JOINERS = SPINEL_PROP_THREAD_EXT__BEGIN + 15,
+
+    /// Thread commissioner enable
+    /** Format `b`
+    *
+    * Default value is `false`.
+    */
+    SPINEL_PROP_THREAD_COMMISSIONER_ENABLED = SPINEL_PROP_THREAD_EXT__BEGIN + 16,
     SPINEL_PROP_THREAD_EXT__END        = 0x1600,
 
     SPINEL_PROP_IPV6__BEGIN          = 0x60,
