--- a/MQTTSNGateway/src/linux/Network.cpp
+++ b/MQTTSNGateway/src/linux/Network.cpp
@@ -21,7 +21,6 @@
 #include <unistd.h>
 #include <fcntl.h>
 #include <sys/socket.h>
-#include <error.h>
 #include <regex>
 
 #include "Network.h"
--- a/MQTTSNGateway/src/linux/Timer.h
+++ b/MQTTSNGateway/src/linux/Timer.h
@@ -17,7 +17,7 @@
 #define MQTTSNGATEWAY_SRC_LINUX_TIMER_H_
 
 #include "MQTTSNGWDefines.h"
-#include <time.h>
+#include <sys/time.h>
 namespace MQTTSNGW
 {
 /*==========================================================
